mapboxgl/templates/choropleth_alt.html

```diff
@@ -0,0 +1,204 @@
+{% extends "main.html" %}
+
+<!-- update legend item key style -->
+{% block extra_css %}
+<style type='text/css'>
+    .legend div span { border-radius: 20%; }
+</style>
+{% endblock extra_css %}
+
+{% block javascript %}
+var legend = document.getElementById('legend');
+
+mapboxgl.accessToken = '{{ accessToken }}';
+
+var map = new mapboxgl.Map({
+    container: 'map',
+    style: '{{ styleUrl }}', 
+    center: {{ center }},
+    zoom: {{ zoom }},
+    transformRequest: (url, resourceType)=> {
+        if ( url.slice(0,22) == 'https://api.mapbox.com' ) {
+            //Add Python Plugin identifier for Mapbox API traffic
+            return {
+                url: [url.slice(0, url.indexOf("?")+1), "pluginName=PythonMapboxgl&", url.slice(url.indexOf("?")+1)].join('')
+            }
+        }
+        else {
+            //Do not transform URL for non Mapbox GET requests
+            return {url: url}
+        }
+    }
+});
+
+var joinData = {{ joinData }};
+
+map.addControl(new mapboxgl.NavigationControl());
+
+calcCircleColorLegend({{ colorStops }}, "{{ colorProperty }}");
+
+function calcNumericColorValue(num, arr) {
+
+    var curr = arr[0];
+    var diff = Math.abs (num - curr);
+    for (var val = 0; val < arr.length; val++) {
+        var newdiff = Math.abs (num - arr[val]);
+        if (newdiff < diff) {
+            diff = newdiff;
+            curr = arr[val];
+        }
+    }
+    return ({{colorStops}})[curr];
+
+}
+
+map.on('style.load', function() {
+    
+    {% if joinData %}
+        var vectorColorStops = [],
+            popUpKeys = [];
+
+        // create categorical color stops for features in vector layer
+        joinData.forEach(function(row, index) {
+            var red = Math.max(255 - row["{{ colorProperty }}"], 0),
```

Comment:
@akacarlyann We should try to generate the data-join match expression in the Python viz class here, instead of in Javascript. This should allow us to re-use the color palette helper functions in the mapboxgl.Colors class, and limit the computation on the javascript side to just loading the join data to create popup values for each property you join from your data set to the polygon feature.

Reply:
Agreed, but I still haven't worked out a way that would exactly replicate the color assignment you would get if the map were just styling points with the same values using an interpolate function.

Reply:
Ah.  I recommend chroma.js for color interpolation -> https://github.com/gka/chroma.js/.  This should be the equivalent in Python for color interpolation -> https://pypi.python.org/pypi/Chroma/0.2.0

I'd prefer if we handle this on the Python side to separate business logic concers to that code base, as opposed to the Jinja template JS files.  The rough steps are:

1. Create a chroma color scale from a color list and domain of stops
2. Calculate the value of color to use for each unique stop
3. Create a mapboxgl `match` expression with the `id` of each feature and `color` calculated from the scale of each color.

Does that make sense?  I think we could wrap up all of 1->3 in one or two helper functions in `mapboxgl.utils`.

Reply:
Can these helper functions be part of a different PR?